File: src/control_package/src/test_motor.cpp
#include <iostream>
#include <lgpio.h>
#include <unistd.h> // for sleep

int main() {
    // ---- CONFIGURATION ----
    // Change these to the actual GPIO numbers you wired up on your Pi 5
    int PWM_PIN = 12;  // The pin connected to PWMA or PWMB (Speed)
    int IN1_PIN = 23;  // The pin connected to AIN1 or BIN1 (Direction 1)
    int IN2_PIN = 24;  // The pin connected to AIN2 or BIN2 (Direction 2)
    
    std::cout << "Initializing lgpio for Raspberry Pi 5..." << std::endl;

    // Pi 5 Specifics: The main 40-pin header is handled by the RP1 chip, 
    // which Linux usually maps to GPIO chip 4. 
    int handle = lgGpiochipOpen(4);
    
    if (handle < 0) {
        // If chip 4 fails, try 0 just in case.
        handle = lgGpiochipOpen(0);
        if (handle < 0) {
            std::cerr << "CRITICAL ERROR: Failed to open GPIO chip." << std::endl;
            std::cerr << "Make sure you installed liblgpio-dev and are using sudo!" << std::endl;
            return 1;
        }
    }

    std::cout << "GPIO Chip Opened successfully." << std::endl;

    // Set the Direction Pins to Output Mode. For Forward: IN1 = 1, IN2 = 0
    if (lgGpioClaimOutput(handle, LG_SET_PULL_NONE, IN1_PIN, 1) < 0) {
        std::cerr << "Error claiming IN1 Pin " << IN1_PIN << std::endl;
        return 1;
    }
    if (lgGpioClaimOutput(handle, LG_SET_PULL_NONE, IN2_PIN, 0) < 0) {
        std::cerr << "Error claiming IN2 Pin " << IN2_PIN << std::endl;
        return 1;
    }

    std::cout << "Spinning Motor at 50% Speed for 3 seconds..." << std::endl;

    // Start PWM on the PWM Pin.
    // Arguments: Handle, Pin, Frequency (Hz), Duty Cycle (0.0 to 100.0)
    if (lgTxPwm(handle, PWM_PIN, 1000, 50.0, 0, 0) < 0) {
         std::cerr << "Error starting PWM on Pin " << PWM_PIN << std::endl;
    }

    // Let the motor run for 3 seconds
    sleep(3);

    std::cout << "Stopping Motor..." << std::endl;
    
    // Stop the motor by setting Duty Cycle to 0.0
    lgTxPwm(handle, PWM_PIN, 1000, 0.0, 0, 0);

    // Free the pins and close the chip safely to prevent locking them out
    lgGpioFree(handle, PWM_PIN);
    lgGpioFree(handle, IN1_PIN);
    lgGpioFree(handle, IN2_PIN);
    lgGpiochipClose(handle);

    std::cout << "Test Complete." << std::endl;
    return 0;
}

File: test_motor.py
import lgpio
import time

# --- CONFIGURATION (TB6612FNG) ---
# Left Motor
PWM_A = 12
AIN1 = 23
AIN2 = 24

# Right Motor
PWM_B = 13
BIN1 = 25
BIN2 = 26

def setup():
    print("Opening GPIO chip...")
    # Open Pi 5's main GPIO chip (usually chip 4)
    try:
        h = lgpio.gpiochip_open(4)
    except:
        h = lgpio.gpiochip_open(0) # Fallback

    # Set as outputs
    for pin in [AIN1, AIN2, BIN1, BIN2]:
        lgpio.gpio_claim_output(h, pin)

    return h

def set_motors(h, left_speed, right_speed):
    # left_speed and right_speed from -100 to 100

    # Left Direction
    if left_speed > 0:
        lgpio.gpio_write(h, AIN1, 1)
        lgpio.gpio_write(h, AIN2, 0)
    elif left_speed < 0:
        lgpio.gpio_write(h, AIN1, 0)
        lgpio.gpio_write(h, AIN2, 1)
    else:
        lgpio.gpio_write(h, AIN1, 0)
        lgpio.gpio_write(h, AIN2, 0)

    # Right Direction
    if right_speed > 0:
        lgpio.gpio_write(h, BIN1, 1)
        lgpio.gpio_write(h, BIN2, 0)
    elif right_speed < 0:
        lgpio.gpio_write(h, BIN1, 0)
        lgpio.gpio_write(h, BIN2, 1)
    else:
        lgpio.gpio_write(h, BIN1, 0)
        lgpio.gpio_write(h, BIN2, 0)

    # Convert to pure positive PWM Duty Cycle
    lpwm = abs(left_speed)
    rpwm = abs(right_speed)

    # Start PWM at 1000 Hz
    lgpio.tx_pwm(h, PWM_A, 1000, lpwm)
    lgpio.tx_pwm(h, PWM_B, 1000, rpwm)

def cleanup(h):
    lgpio.tx_pwm(h, PWM_A, 1000, 0)
    lgpio.tx_pwm(h, PWM_B, 1000, 0)
    for pin in [PWM_A, AIN1, AIN2, PWM_B, BIN1, BIN2]:
        lgpio.gpio_free(h, pin)
    lgpio.gpiochip_close(h)

if __name__ == "__main__":
    h = setup()
    try:
        print("Starting Forward Loop...")
        # Ramp up speed
        for speed in range(0, 101, 10):
            print(f"Speed: {speed}%")
            set_motors(h, speed, speed)
            time.sleep(0.5)

        # Ramp down
        for speed in range(100, -1, -10):
            print(f"Speed: {speed}%")
            set_motors(h, speed, speed)
            time.sleep(0.5)

        print("Starting Reverse Loop...")
        # Ramp up reverse
        for speed in range(0, -101, -10):
            print(f"Speed: {speed}%")
            set_motors(h, speed, speed)
            time.sleep(0.5)

        # Ramp down reverse
        for speed in range(-100, 1, 10):
            print(f"Speed: {speed}%")
            set_motors(h, speed, speed)
            time.sleep(0.5)

        print("Test Complete!")

    except KeyboardInterrupt:
        print("\nManually stopped.")
    finally:
        cleanup(h)
